internal/nmc/module_nmc_controller.go

```diff
@@ -0,0 +1,142 @@
+		} else {
+			err = mnh.reconHelper.disableModuleOnNode(ctx, mld.Namespace, mld.Name, node.Name)
+		}
+
+		sumErr = errors.Join(sumErr, err)
+	}
+
+	if sumErr != nil {
+		return fmt.Errorf("failed to handle module %s/%s with some of the nodes: %v", mld.Namespace, mld.Name, sumErr)
+	}
+	return nil
+}
+
+//go:generate mockgen -source=module_nmc_controller.go -package=nmc -destination=mock_module_nmc_controller.go
+
+type moduleNMCHandlerHelperAPI interface {
+	shouldModuleRunOnNode(node v1.Node, mld *api.ModuleLoaderData, nodeKernelVersion string) (bool, error)
+	enableModuleOnNode(ctx context.Context, mld *api.ModuleLoaderData, nodeName, kernelVersion string) error
+	disableModuleOnNode(ctx context.Context, modNamespace, modName, nodeName string) error
+}
+
+type moduleNMCHandlerHelper struct {
+	client    client.Client
+	nmcHelper Helper
+}
+
+func newModuleNMCHandlerHelper(client client.Client, nmcHelper Helper) moduleNMCHandlerHelperAPI {
+	return &moduleNMCHandlerHelper{
+		client:    client,
+		nmcHelper: nmcHelper,
+	}
+}
+
+func (mnhh *moduleNMCHandlerHelper) shouldModuleRunOnNode(node v1.Node, mld *api.ModuleLoaderData, nodeKernelVersion string) (bool, error) {
+	if nodeKernelVersion != mld.KernelVersion {
+		return false, nil
+	}
+
+	if !utils.IsNodeSchedulable(&node) {
+		return false, nil
+	}
+
+	return utils.IsObjectSelectedByLabels(node.GetLabels(), mld.Selector)
+}
+
+func (mnhh *moduleNMCHandlerHelper) enableModuleOnNode(ctx context.Context, mld *api.ModuleLoaderData, nodeName, kernelVersion string) error {
+	logger := log.FromContext(ctx)
+	moduleConfig := kmmv1beta1.ModuleConfig{
+		KernelVersion:        kernelVersion,
+		ContainerImage:       mld.ContainerImage,
+		InTreeModuleToRemove: mld.InTreeModuleToRemove,
+		Modprobe:             mld.Modprobe,
+	}
+
+	nmc := &kmmv1beta1.NodeModulesConfig{
+		ObjectMeta: metav1.ObjectMeta{Name: nodeName},
+	}
+
+	opRes, err := controllerutil.CreateOrPatch(ctx, mnhh.client, nmc, func() error {
+		return mnhh.nmcHelper.SetNMCAsDesired(ctx, nmc, mld.Namespace, mld.Name, &moduleConfig)
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to enable module %s/%s in NMC %s: %v", mld.Namespace, mld.Name, nodeName, err)
+	}
+	logger.Info("enable module in NMC", "name", mld.Name, "namespace", mld.Namespace, "node", nodeName, "result", opRes)
+	return nil
+}
+
+func (mnhh *moduleNMCHandlerHelper) disableModuleOnNode(ctx context.Context, modNamespace, modName, nodeName string) error {
+	logger := log.FromContext(ctx)
+	nmc, err := mnhh.nmcHelper.Get(ctx, nodeName)
+	if err != nil {
+		return fmt.Errorf("failed to get the NodeModulesConfig for node %s: %v", nodeName, err)
+	}
+	if nmc == nil {
+		// NodeModulesConfig does not exists, module was never running on the node, we are good
+		return nil
+	}
+
+	opRes, err := controllerutil.CreateOrPatch(ctx, mnhh.client, nmc, func() error {
+		return mnhh.nmcHelper.SetNMCAsDesired(ctx, nmc, modNamespace, modName, nil)
```

Comment:
Reading this code again, I think it would be more straightforward with a dedicated function:
```go
func RemoveModuleConfig(nmsList *[]NodeModuleSpec*, nmsNamespace, nmsName string)
```
This would very much look like the [conditions management code](https://github.com/kubernetes/apimachinery/blob/5cb236977966cfc829e3b94b57fa69b44e0a95bc/pkg/api/meta/conditions.go#L57-L71).
WDYT?

Reply:
Fixed